File: src/infusion_explorer.py
"""Static shell and reviewed data for the infusion concentration explorer."""
import json
from html import escape
from pathlib import Path


def load_levels():
    return json.loads(Path(__file__).with_name('infusion-levels.json').read_text(encoding='utf-8'))


def level_label(row):
    low, high = row.get('bandLow'), row.get('bandHigh')
    if low is not None and high is not None:
        return f"{low:g}–{high:g} {row['unit']}"
    if high is not None:
        return f"Caution boundary: {high:g} {row['unit']}"
    return 'No serum target in this model'


def render_infusion_explorer():
    data = load_levels()
    options = ''.join(f'<option value="{escape(r["id"])}">{escape(r["medication"])}</option>' for r in data['rows'])
    payload = json.dumps(data, ensure_ascii=False).replace('<', '\\u003c')
    return f'''<div class="infusion-explorer" id="infusion-explorer" data-infusion-explorer>
<div class="explorer-heading"><div><h3>Explore delivery and concentration</h3><p>Choose a medication, change its actual delivered rate and inspect how its game value accumulates or clears.</p></div><a href="#medication-mixtures" class="explorer-link">Prepare a mixture &gt;</a></div>
<p class="infusion-mode-note">{escape(data["modeNote"])} <a href="#slow-medication-pushes">Manual push guidance &gt;</a></p>
<div class="explorer-controls" data-explorer-controls hidden>
<label class="explorer-drug"><span>Medication</span><select data-level-drug>{options}</select></label>
<label data-model-control><span>Delivered rate <span data-rate-unit>(mg/min)</span></span><input data-level-rate type="number" min="0" step="any" value="15" inputmode="decimal"></label>
<label data-model-control><span>Chart duration (min)</span><select data-level-duration><option value="15">15</option><option value="30">30</option><option value="60" selected>60</option><option value="120">120</option><option value="240">240</option><option value="360">360</option></select></label>
<label data-weight-control hidden><span>Patient weight (kg)</span><input data-level-weight type="number" min="30" max="200" step="1" value="83" inputmode="decimal"></label>
<div class="explorer-rate-track" data-model-control><label for="level-rate-slider">Adjust delivered rate</label><input id="level-rate-slider" data-level-rate-slider type="range" min="0" max="120" step="0.1" value="15"><span>Slider limits are examples, not dose limits.</span></div>
<div class="explorer-stop" data-model-control><label class="explorer-check"><input data-level-stop-enabled type="checkbox"><span>Stop delivery during the chart</span></label><label><span>Stop at (min)</span><input data-level-stop type="number" min="0" max="60" step="1" value="30" disabled></label></div>
<fieldset class="explorer-modifiers" data-lidocaine-controls hidden><legend>Lidocaine clearance modifiers</legend><label class="explorer-check"><input type="checkbox" data-level-shock><span>Shock modifier (0.6× clearance)</span></label><label class="explorer-check"><input type="checkbox" data-level-esmolol><span>Esmolol proxy &gt; 0.1 (0.7× clearance)</span></label></fieldset>
</div>
<p class="ref-error" data-level-error hidden role="status"></p>
<div data-level-output hidden>
<div class="explorer-metrics"><div><h4 data-level-reference-label>Configured reference</h4><strong data-level-band></strong><span data-level-band-note></span></div><div><h4 data-level-current-label>At the selected time</h4><strong data-level-current></strong><span data-level-time></span></div><div><h4 data-level-plateau-label>With continuous delivery</h4><strong data-level-plateau></strong><span data-level-half-life></span></div></div>
<div class="explorer-plot" data-level-plot>
<p class="explorer-unit" data-level-axis></p>
<svg data-level-svg viewBox="0 0 780 270" role="img" aria-labelledby="level-chart-title level-chart-desc"><title id="level-chart-title">Estimated game concentration over time</title><desc id="level-chart-desc">Use the time slider or touch the curve for exact values. Shading shows the configured debug reference, where one exists.</desc><g data-level-grid></g><rect data-level-shade fill="#f1bd59" opacity="0.10"></rect><g data-level-boundaries></g><path data-level-path fill="none" stroke="#55baf3" stroke-width="3" stroke-linejoin="round"></path><line data-level-cursor stroke="#a8b7c3" stroke-dasharray="4 5" y1="14" y2="226"></line><circle data-level-dot r="5" fill="#f1bd59" stroke="#071015" stroke-width="2"></circle><rect x="60" y="14" width="704" height="212" fill="transparent" data-level-hit></rect></svg>
<div class="explorer-legend"><span><i class="legend-line"></i>Estimated game value</span><span data-level-shade-key><i class="legend-band"></i>Configured reference</span><span>Time in minutes</span></div>
<label class="explorer-time"><span>Inspect time <output data-level-cursor-time>60 min</output></span><input data-level-time-slider type="range" min="0" max="60" step="0.1" value="60"></label>
<p class="explorer-readout" data-level-readout aria-live="polite" aria-atomic="true"></p>
</div>
<p class="explorer-meaning" data-level-meaning></p>
<ul class="explorer-boundaries" data-level-thresholds></ul>
<a data-level-card-link href="#infusion-risk-amiodarone">Dose ranges and cautions &gt;</a>
</div>
<noscript><p>Enable JavaScript for the concentration explorer. Every medication's reference band, timing and cautions remain available in the expandable guide below.</p></noscript>
<details class="explorer-assumptions"><summary>What these numbers can tell you</summary><div><p><strong>There is no universal ideal serum level.</strong> Most bands here are configured debug references. Sedation, pressor response and overdose may read different variables. Lidocaine, esmolol and calcium have explicit concentration or excess thresholds; those boundaries are not a guarantee of safety below them.</p><p>The curve starts at zero, holds actual admitted delivery constant and uses the default configuration. It estimates the source's accumulation in one second steps, including calcium and amiodarone input easing. A prior bolus, changing perfusion, lost IV fluid, addon settings or another drug can change the live result. Calcium is an excess accumulator shared by both salts, not a laboratory calcium measurement. Magnesium's debug band does not align with its raw proxy scale at native treatment rates.</p><p><strong>IV and IO:</strong> the same amount admitted to circulation feeds this estimate. Check the site's actual delivery; a leaking IV can drain a bag without admitting that entire dose.</p></div></details>
<p class="source-note">Concentration and mixture review: <a href="https://github.com/{data['repository']}/tree/{data['revision']}">ACM Extended dev {data['revision'][:7]}</a>. Manual push comparison: <a href="https://github.com/{data['repository']}/blob/{data['rateReviewRevision']}/addons/acm_extended/functions/fn_medicationDriveAdd.sqf">{data['rateReviewRevision'][:7]}</a>. Individual source links accompany each medication below.</p>
<script type="application/json" id="infusion-level-data">{payload}</script>
</div>'''

File: src/infusion-explorer.js
/* Game estimates only. The source supplies bands, units and accumulation rules. */
(() => {
  const root = document.querySelector('[data-infusion-explorer]');
  if (!root) return;
  const data = JSON.parse(document.getElementById('infusion-level-data').textContent);
  const $ = name => root.querySelector(`[data-${name}]`);
  const finite = value => Number.isFinite(Number(value));
  const fmt = value => Number(value).toLocaleString('en-US', { maximumSignificantDigits: 4 });
  const scale = row => row.rateUnit === 'mcg/min' ? 1000 : 1;
  const label = row => row.bandLow != null ? `${fmt(row.bandLow)}–${fmt(row.bandHigh)} ${row.unit}`
    : row.bandHigh != null ? `${fmt(row.bandHigh)} ${row.unit}` : 'No serum target';

  function project(row, options) {
    const { rateMgMin, minutes, weightKg = 83, stopMinutes = null, shock = false, esmolol = false } = options;
    if (row.model === 'none') return null;
    if (![rateMgMin, minutes, weightKg].every(finite) || rateMgMin < 0 || rateMgMin > 1000000 || minutes <= 0 || minutes > 360 || weightKg < 30 || weightKg > 200 || (stopMinutes != null && (!finite(stopMinutes) || stopMinutes < 0 || stopMinutes > minutes))) throw new RangeError('Invalid infusion inputs');
    let clearance = row.clearance;
    if (row.id === 'lidocaine') clearance = Math.max(0.05, clearance * (shock ? 0.6 : 1) * (esmolol ? 0.7 : 1));
    const k = row.model === 'one-compartment' ? clearance / (60 * weightKg * row.volumeLPerKg) : 0.693 / row.halfLifeSec;
    const plateau = rateMgMin * row.inputFactor / clearance;
    const points = [{ seconds: 0, level: 0, deliveredMg: 0 }];
    let level = 0, easedRate = 0, deliveredMg = 0;
    const end = Math.round(minutes * 60), stop = stopMinutes == null ? Infinity : stopMinutes * 60;
    for (let t = 1; t <= end; t++) {
      const admittedRate = t - 1 < stop ? rateMgMin : 0;
      if (row.inputEasing) {
        const tau = admittedRate > easedRate ? row.inputEasing.onsetSec : row.inputEasing.offsetSec;
        easedRate += (admittedRate - easedRate) * Math.min(1 / tau, 1);
      } else easedRate = admittedRate;
      const target = easedRate * row.inputFactor / clearance;
      level += k * (target - level);
      deliveredMg += admittedRate / 60;
      points.push({ seconds: t, level: Math.max(0, level), deliveredMg });
    }
    return { points, plateau, halfLifeMinutes: Math.log(2) / k / 60, clearance };
  }
  // Small public math surface lets verification check against independent values.
  window.ACMInfusionModel = Object.freeze({ project });
  let row = data.rows[0], result, yMax = 1, chartWidth = 780;
  const select = $('level-drug'), rate = $('level-rate'), slider = $('level-rate-slider');
  const duration = $('level-duration'), stopEnabled = $('level-stop-enabled'), stop = $('level-stop');
  const weight = $('level-weight'), time = $('level-time-slider');
  const svgNS = 'http://www.w3.org/2000/svg';
  const el = (tag, attrs, value) => {
    const node = document.createElementNS(svgNS, tag);
    Object.entries(attrs).forEach(([key, val]) => node.setAttribute(key, val));
    if (value != null) node.textContent = value;
    return node;
  };
  const x = minute => 60 + (chartWidth - 76) * minute / Number(duration.value);
  const y = value => 226 - 212 * value / yMax;
  const txt = (name, value) => { $(name).textContent = value; };
  function thresholds(point) {
    const list = $('level-thresholds');
    list.replaceChildren();
    for (const boundary of row.thresholds) {
      const item = document.createElement('li'), value = document.createElement('b');
      value.textContent = `${boundary.comparison} ${fmt(boundary.value)} ${row.unit}: `;
      item.append(value, `${boundary.label}${boundary.detail ? '. ' + boundary.detail : ''}`);
      if (point && boundary.purpose === 'toxicity') item.dataset.reached = String(boundary.comparison === '>' ? point.level > boundary.value : point.level >= boundary.value);
      list.append(item);
    }
    for (const boundary of row.rateThresholds) {
      const item = document.createElement('li'), value = document.createElement('b');
      value.textContent = `${boundary.comparison} ${fmt(boundary.value * scale(row))} ${row.rateUnit}: `;
      item.append(value, `${boundary.label}. Reads ${boundary.basis || 'actual admitted rate'}.`);
      list.append(item);
    }
    list.hidden = list.children.length === 0;
  }
  function inspect(minute) {
    if (!result) return;
    const point = result.points[Math.min(result.points.length - 1, Math.max(0, Math.round(minute * 60)))];
    const actualMinute = point.seconds / 60;
    time.value = actualMinute;
    txt('level-cursor-time', `${fmt(actualMinute)} min`);
    txt('level-current', `${fmt(point.level)} ${row.unit}`);
    txt('level-time', `At ${fmt(actualMinute)} min from a zero starting value`);
    const dose = row.rateUnit === 'mcg/min' ? `${fmt(point.deliveredMg * 1000)} mcg` : `${fmt(point.deliveredMg)} mg`;
    txt('level-readout', `${fmt(actualMinute)} min · ${fmt(point.level)} ${row.unit} · ${dose} admitted`);
    $('level-dot').setAttribute('cx', x(actualMinute));
    $('level-dot').setAttribute('cy', y(point.level));
    window.ACMEWiki.sizeChartMarkers($('level-svg'), [$('level-dot')]);
    $('level-cursor').setAttribute('x1', x(actualMinute));
    $('level-cursor').setAttribute('x2', x(actualMinute));
    thresholds(point);
  }
  function draw() {
    chartWidth = Math.max(140, $('level-svg').clientWidth);
    $('level-svg').setAttribute('viewBox', `0 0 ${chartWidth} 270`);
    $('level-hit').setAttribute('width', chartWidth - 76);
    const peak = Math.max(...result.points.map(p => p.level));
    yMax = Math.max(peak * 1.15, (row.bandHigh || 0) * 1.5, 0.0001);
    const grid = $('level-grid');
    grid.replaceChildren();
    for (let i = 0; i <= 4; i++) {
      const value = yMax * i / 4, minute = Number(duration.value) * i / 4;
      grid.append(el('line', { x1: 60, x2: chartWidth - 16, y1: y(value), y2: y(value), stroke: '#203440', 'stroke-width': 1 }));
      grid.append(el('text', { x: 52, y: y(value) + 4, 'text-anchor': 'end' }, fmt(value)));
      if (chartWidth > 400 || i % 2 === 0) grid.append(el('text', { x: x(minute), y: 249, 'text-anchor': i === 4 ? 'end' : 'middle' }, fmt(minute)));
    }
    const band = $('level-shade'), bounds = $('level-boundaries');
    bounds.replaceChildren();
    const shaded = row.bandLow != null && row.bandHigh != null;
    band.style.display = shaded ? '' : 'none';
    $('level-shade-key').hidden = !shaded;
    if (shaded) {
      Object.entries({ x: 60, y: y(row.bandHigh), width: chartWidth - 76, height: y(row.bandLow) - y(row.bandHigh) }).forEach(([key, value]) => band.setAttribute(key, value));
    }
    const boundaries = [...new Set([row.bandLow, row.bandHigh, ...row.thresholds.filter(b => b.purpose === 'toxicity').map(b => b.value)].filter(v => v != null && v <= yMax))];
    boundaries.forEach(value => bounds.append(el('line', { x1: 60, x2: chartWidth - 16, y1: y(value), y2: y(value), stroke: '#cfaa65', opacity: .6, 'stroke-dasharray': '5 6' })));
    const step = Math.max(1, Math.floor(result.points.length / 700));
    const points = result.points.filter((p, i) => i % step === 0 || i === result.points.length - 1);
    $('level-path').setAttribute('d', points.map((p, i) => `${i ? 'L' : 'M'}${x(p.seconds / 60).toFixed(2)},${y(p.level).toFixed(2)}`).join(' '));
    document.getElementById('level-chart-title').textContent = `${row.medication}: estimated game concentration`;
    txt('level-axis', row.unit);
  }
  function render() {
    txt('level-meaning', row.meaning);
    txt('level-band', label(row));
    txt('level-reference-label', row.bandLow == null && row.bandHigh != null ? 'Caution boundary' : 'Configured reference');
    txt('level-band-note', row.therapeuticDisplayLabel);
    $('level-card-link').href = `#infusion-risk-${row.id}`;
    const modeled = row.model !== 'none';
    root.querySelectorAll('[data-model-control]').forEach(control => { control.hidden = !modeled; });
    $('weight-control').hidden = row.model !== 'one-compartment';
    $('lidocaine-controls').hidden = row.id !== 'lidocaine';
    $('level-plot').hidden = !modeled;
    $('level-output').hidden = false;
    $('level-error').hidden = true;
    if (!modeled) {
      result = null;
      txt('level-current-label', row.osmotherapy ? 'Modeled ICP response' : 'What the game reads');
      txt('level-current', row.osmotherapy ? `Up to ${row.osmotherapy.icpDropPerReferenceMmHg} mmHg` : 'Native medication effect');
      txt('level-time', row.osmotherapy ? `Per ${row.osmotherapy.referenceVolumeMl} mL reference with a TBI state` : 'Medication already on board, rather than a serum target');
      txt('level-plateau-label', row.osmotherapy ? 'Sodium ceiling' : 'Preparation');
      txt('level-plateau', row.osmotherapy ? `${row.osmotherapy.sodiumCeiling} mEq/L` : 'Single drug or mixture');
      txt('level-half-life', row.osmotherapy ? 'Benefit stops here; this is not a target.' : 'See Ketofol and the dose guide for combined effects.');
      thresholds(null);
      return;
    }
    txt('level-current-label', 'At the selected time');
    txt('level-plateau-label', 'With continuous delivery');
    const rateValue = rate.value === '' ? NaN : Number(rate.value);
    const options = { rateMgMin: rateValue / scale(row), minutes: Number(duration.value), weightKg: row.model === 'one-compartment' ? (weight.value === '' ? NaN : Number(weight.value)) : 83, stopMinutes: stopEnabled.checked ? (stop.value === '' ? NaN : Number(stop.value)) : null, shock: $('level-shock').checked, esmolol: $('level-esmolol').checked };
    try { result = project(row, options); } catch {
      result = null;
      txt('level-error', rateValue / scale(row) > 1000000 ? 'That value exceeds this calculator’s numerical range. Enter a smaller value.' : 'Enter a nonnegative delivered rate, a weight of 30–200 kg and a stop time inside the chart duration.');
      $('level-error').hidden = false;
      $('level-output').hidden = true;
      return;
    }
    txt('level-plateau', `${fmt(result.plateau)} ${row.unit}`);
    txt('level-half-life', `Estimated plateau if delivery continues; washout half-life about ${fmt(result.halfLifeMinutes)} min.`);
    const previousTime = Number(time.value), atEnd = previousTime === Number(time.max);
    time.max = duration.value;
    time.value = atEnd ? duration.value : Math.min(previousTime, Number(duration.value));
    stop.max = duration.value;
    slider.value = Math.min(rateValue, Number(slider.max));
    draw();
    inspect(Number(time.value));
  }
  function choose(id) {
    row = data.rows.find(item => item.id === id) || data.rows[0];
    select.value = row.id;
    if (row.model !== 'none') {
      rate.value = row.defaultRateMgMin * scale(row);
      slider.max = row.rateMaxMgMin * scale(row);
      slider.step = Number(slider.max) / 1000;
      txt('rate-unit', `(${row.rateUnit})`);
    }
    render();
  }
  select.addEventListener('change', () => choose(select.value));
  for (const control of [rate, weight, stop, $('level-shock'), $('level-esmolol')]) control.addEventListener('input', render);
  duration.addEventListener('input', () => { stop.value = Math.min(Number(stop.value), Number(duration.value)); render(); });
  slider.addEventListener('input', () => { rate.value = slider.value; render(); });
  stopEnabled.addEventListener('change', () => { stop.disabled = !stopEnabled.checked; render(); });
  time.addEventListener('input', () => inspect(Number(time.value)));
  const hit = $('level-hit');
  function pointer(event) {
    const matrix = $('level-svg').getScreenCTM();
    if (!matrix || !result) return;
    const point = new DOMPoint(event.clientX, event.clientY).matrixTransform(matrix.inverse());
    inspect(Math.max(0, Math.min(1, (point.x - 60) / (chartWidth - 76))) * Number(duration.value));
  }
  hit.addEventListener('pointermove', event => { if (event.pointerType === 'mouse' || event.buttons) pointer(event); });
  hit.addEventListener('pointerdown', pointer);
  document.querySelectorAll('[data-explore-infusion]').forEach(button => {
    button.hidden = false;
    button.addEventListener('click', () => {
      choose(button.dataset.exploreInfusion);
      root.scrollIntoView({ behavior: matchMedia('(prefers-reduced-motion: reduce)').matches ? 'auto' : 'smooth', block: 'start' });
      select.focus({ preventScroll: true });
    });
  });
  $('explorer-controls').hidden = false;
  choose(select.value);
  new ResizeObserver(() => {
    if (result && Math.abs($('level-svg').clientWidth - chartWidth) > 1) { draw(); inspect(Number(time.value)); }
  }).observe($('level-svg'));
})();
